fuse/nodefs/fsconnector.go

```diff
@@ -407,6 +407,26 @@ func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data
 	return c.server.InodeNotifyStoreCache(nID, off, data)
 }
 
+// FileRetrieveCache retrieves data from kernel's inode cache.
+//
+// This call retrieves data from kernel's inode cache @ offset and up to
+// len(dest) bytes. If kernel cache has fewer consequtive data starting at
+// offset, that fewer amount is returned. In particular if inode data at offset
+// is not cached (0, OK) is returned.
```

Comment:
Hm, we have this pattern in many places in nodefs: when a function accepts an Inode and tries to convert the Inode into integer node ID, and if it does not find the handle for the Inode it returns EINVAL. For example here:

https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L422
https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L440
https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L404

and none of this functions mention EINVAL in their docstring.

Actually isn't it a _bug_, if `Inode` is used but is not in the handle map?
Shouldn't in that case it thus should be not EINVAL, but panic instead?

I agree that similar functions at raw fuse level might return EINVAL abour wrong node ID, but that EINVAL will come from kernel in that case. However at nodefs level there should not be case when a valid Inode is used and it is not in `inodeMap`. Or am I missing something?

Given the above what we should do for my `nodefs.FileRetrieveCache(Inode, ...)`? Document that if invalid Inode is passed in EINVAL will be returned? or we should just conver all the above cases to panics?

P.S. 

https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L381 returns even OK if the Inode was not found in inodeMap. The rationale here is, probably, that since there is nothing in inode map, the kernel does not know about this inode, and then there is no need to notify it about that the inode content has been changed. But still, isn't it a bug?

Reply:
if there is no inode number, we can't talk to the kernel, so I think there is nothing to be done. EINVAL isn't documented in other places, but maybe those other places should be fixed?

The other Notify functions can return OK, because if the kernel doesn't know about the inode, then its cache can certainly not be out of date.

Reply:
Ok, thanks for clarifying. I also thought a bit and looked at fuse inside kernel a bit and now it all makes sense: if the kernel did not lookup on the inode, or it did the lookup but later forgot, it does not have an entry for the inode in its dentry cache and so directly talking to kernel with any notify operation about the inode will return `ENODEV`.

Now for operations that have semantic to just invalidate the cache, we can pretend that they succeed. For `FileRetrieveCache` we can also pretend that the cache is empty and reads as just (0, OK) instead of EINVAL.

For `FileNotifyStoreCache` I indeed have to fix the docstring and the implementation to return ENOENT for the case when the kernel does not have a handle for the inode. And it was the only case where we were returning EINVAL.

I will amend the whole change with the following:

```diff
diff --git a/fuse/nodefs/fsconnector.go b/fuse/nodefs/fsconnector.go
index 5010dd8..2a94319 100644
--- a/fuse/nodefs/fsconnector.go
+++ b/fuse/nodefs/fsconnector.go
@@ -393,6 +393,9 @@ func (c *FileSystemConnector) FileNotify(node *Inode, off int64, length int64) f
 // and will use data from that cache for non direct-IO reads from the inode
 // in corresponding data region. After kernel's cache data is evicted, the kernel
 // will have to issue new Read calls on user request to get data content.
+//
+// ENOENT is returned if the kernel does not currently have entry for this
+// inode in its dentry cache.
 func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data []byte) fuse.Status {
        var nID uint64
        if node == c.rootNode {
@@ -402,7 +405,8 @@ func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data
        }
 
        if nID == 0 {
-               return fuse.EINVAL
+               // the kernel does not currently know about this inode.
+               return fuse.ENOENT
        }
        return c.server.InodeNotifyStoreCache(nID, off, data)
 }
@@ -422,7 +426,9 @@ func (c *FileSystemConnector) FileRetrieveCache(node *Inode, off int64, dest []b
        }
 
        if nID == 0 {
-               return 0, fuse.EINVAL
+               // the kernel does not currently know about this inode.
+               // -> we can pretend that its cache for the inode is empty.
+               return 0, fuse.OK
        }
        return c.server.InodeRetrieveCache(nID, off, dest)
 }
diff --git a/fuse/test/cachecontrol_test.go b/fuse/test/cachecontrol_test.go
index 78cba41..5c1bc84 100644
--- a/fuse/test/cachecontrol_test.go
+++ b/fuse/test/cachecontrol_test.go
@@ -121,8 +121,15 @@ func TestCacheControl(t *testing.T) {
                assertCacheReadAt(0, l+10000, dataOK)
        }
 
+       // before the kernel has entry for file in its dentry cache, the cache
+       // should read as empty and store should fail with ENOENT.
+       assertCacheRead("before lookup", "")
+       st := fsconn.FileNotifyStoreCache(file.Inode(), 0, []byte("abc"))
+       if st != fuse.ENOENT {
+               t.Fatalf("%s: store cache -> %v; want %v", "before lookup", st, fuse.ENOENT)
+       }
+
        // lookup on the file - forces to assign inode ID to it
-       // XXX better do it without kernel - how?
        os.Stat(dir + "/hello.txt")
 
        // cache should be initially empty
@@ -168,7 +175,7 @@ func TestCacheControl(t *testing.T) {
        assertCacheRead("original", data0)
 
        // store changed data into OS cache
-       st := fsconn.FileNotifyStoreCache(file.Inode(), 7, []byte("123"))
+       st = fsconn.FileNotifyStoreCache(file.Inode(), 7, []byte("123"))
        if st != fuse.OK {
                t.Fatalf("store cache: %s", st)
        }
```